### PR TITLE
Fixed native versions + Null safety

### PR DESCRIPTION
**Pull request by radvansky-tomas**
My changes with latest null-safety branch

### CODE REVIEW

File: example/ios/Podfile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Uncomment this line to define a global platform for your project
-# platform :ios, '9.0'
+platform :ios, '12.0'
```
Comment (hpoul) [inline example/ios/Podfile:2]: Is there a reason this requires ios 12? (it doesn't really matter as this is just the example, but i guess it should work down to ios 9?)

File: example/ios/Runner.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -168,7 +170,7 @@
 				TargetAttributes = {
 					97C146ED1CF9000F007C117D = {
 						CreatedOnToolsVersion = 7.3.1;
-						DevelopmentTeam = 2N9YTZSQBW;
+						DevelopmentTeam = HB3FQU7HQM;
```
Comment (hpoul) [inline example/ios/Runner.xcodeproj/project.pbxproj:173]: can you remove the changes in this file? Also, is the entitlements file with the `keychain-access-groups` required?

File: macos/Classes/BiometricStorageImpl.swift [hunk 3]
```diff
@@ -30,8 +30,10 @@ class BiometricStorageImpl {
   init(storageError: @escaping StorageError, storageMethodNotImplemented: Any) {
     self.storageError = storageError
     self.storageMethodNotImplemented = storageMethodNotImplemented
+    self.context = LAContext()
   }
   
+private var context:LAContext
```
Comment (hpoul) [inline macos/Classes/BiometricStorageImpl.swift:36]:
```suggestion
  private var context:LAContext
```

**Review by hpoul — CHANGES_REQUESTED** [review]
thanks for the PR. Can you please remove the few unnecessary changes from the ios/macos project? thanks.

I think it's weird that the android api doesn't handle the validity itself. but anyway. I think I'll make this configurable, because i'd rather have the user authenticate every time, but the as far as i understand it, the validity seconds are always required when using this api. but i'll do it after the merge.

**Comment by radvansky-tomas** [comment]
Thank you for comments, will check it tonight. With regards to keychain sharing, that was somewhere mentioned that is required...but will recheck it on physical device.

File: example/ios/Podfile (hunk 1 above)
Comment (radvansky-tomas) [reply example/ios/Podfile:2]: https://developer.apple.com/documentation/localauthentication/logging_a_user_into_your_app_with_face_id_or_touch_id

LAContext is supported from 11.3 I think, I am trying only to support last 2 versions. But as you said this is only example app. Will try to change it to 11.3 and try it on physical device

File: example/ios/Runner.xcodeproj/project.pbxproj (hunk 2 above)
Comment (radvansky-tomas) [reply example/ios/Runner.xcodeproj/project.pbxproj:173]: Will change team, keychain might be required will check that

File: example/ios/Podfile (hunk 1 above)
Comment (radvansky-tomas) [reply example/ios/Podfile:2]: This was incorrect info...LAContext is supported from earlier version, however from 11.3 it allows you to specify what kind of biometry you want to use, what is available, get more error handling and messages...basically more mature version...I will try to move it to iOS10

Comment (hpoul) [reply example/ios/Podfile:2]: according to the documentation it looks like it should be supported in iOS 9: https://developer.apple.com/documentation/security/ksecuseauthenticationcontext

```
iOS 9.0+
macOS 10.11+
```

**Comment by radvansky-tomas** [comment]
So update, I spent hours yesterday trying to make this work...

However something had change since last time I played with it. To give more overview you can read through this https://developer.apple.com/documentation/localauthentication/accessing_keychain_items_with_face_id_or_touch_id#see-also

Basically to use `touchIDAuthenticationAllowableReuseDuration` we have to reuse `LAContext` which I've done. However since last version, behaviour has changed completely.

Before, without reused LAContext, Face/TouchID has been used for every access of keychain, no matter of grace period set. With my reusing of LAContext it worked as expect, but now - it asks for Face/TouchID ONCE ONLY, and then any sub-sequential request is allowed. (like infinite `touchIDAuthenticationAllowableReuseDuration period`)

I already filed report to apple, as same behaviour can be replicated using their example project on native app. I would like to kindly ask anyone who can to try that example project and let me know if they have similar experience...you can use their app how it is...and then even try to reuse LAContext (by placing in on top and then add this line to queries):

` kSecUseAuthenticationContext as String: context,`

Anyhow, till this is resolved I am suggested NOT TO MERGE it.

If this is expected behaviour and I will get confirmation from APPLE that this is new and better way :D then I will implement timer and invalidate context in selected time period (which is really stupid)

**Comment by amelmusic** [comment]
Any news regarding this?

**Comment by MatthewLM** [comment]
I'd be appreciative of the `authenticationValidityDurationSeconds` enhancement (e370c71).

**Comment by MatteoGheza** [comment]
Any update on this?

**Comment by radvansky-tomas** [comment]
@MatteoGheza I will revisit this again probably this weekend and see if new iOS updates fixed anything, or if there is more official documentation about usage of it. If not I will probably contact apple directly, as it bit weird how it was working last time

**Comment by hpoul** [comment]
i'm going to close this PR, because it's probably un-mergeable at this point due to a lot of changes. If anyone is in interested in the `authenticationValidityDurationSeconds` implementation, please create a new PR with only that change. thanks!

**Pull request closed without merging**